### PR TITLE
introduce "random" to the right side of conditions

### PR DESCRIPTION
**Pull request by EndrosG**
Introduce "random" and playerConditions to the right side of conditions, allowing for:
- "random var" <= random
- "random var" = random
- "CoalitionJobsOld" = "CoalitionJobs"
- "reputation: Navy Intelligence" >= "reputation: Republic"

### CODE REVIEW

**Comment by Amazinite** [comment]
>"reputation: Navy Intelligence" >= "reputation: Republic"

Sounds fun,

**Comment by EndrosG** [comment]
yeah, with this PR you should be able to check in missions if the Intelligence likes you more than the normal Republic... or if the Bounty Hunters like you more than the Syndicate Extremists... or whatever comes to your mind.

Actually you could even do some more complicate comparison like:
is reputation: one alien faction > average reputation of other alien faction
by doing
"left side" = "reputation: one alien faction"
"left side" += "reputation: one alien faction"
"left side" += "reputation: one alien faction"

"right side" = "reputation: first other alien faction"
"right side" += "reputation: second other alien faction"
"right side" += "reputation: third other alien faction"
and then finally comparing "left side" > "right side"

And if you then combine that with:
```
	and
		or
			"right side" < "reputation: first other alien faction"
			"right side" < "reputation: second other alien faction"
			"right side" < "reputation: third other alien faction"
		"left side" > "right side"
```
--> one alien says: We know you like us quite a bit, but we would like to be the ones you like most... So, please work even harder for our major interests... or you will loose your license to buy our weaponry...

**Comment by Amazinite** [comment]
*Arbitrary variable useage approaches*

**Comment by EndrosG** [comment]
the official roadmap says:
```
Missions:
    Two more main story lines.
    More side missions.
    Allow arbitrary condition expressions.
```
So... this is IMHO one step on that path.

**Comment by Amazinite** [comment]
Yup.

**Comment by Amazinite** [comment]
{"/=", [](int a, int b) { return a / b; }},
if b == 0, set b to 1 (i.e. nothing occurs)?

**Comment by endless-sky** [comment]
This won't serialize the new conditions, so they'll be lost as soon as you save and reload. And, it would probably make more sense to use whether or not the second string is empty as the switch to control whether the number or the string is used.

Also, you should take a look at the [coding standard](https://endless-sky.github.io/styleguide/styleguide.xml).

**Comment by EndrosG** [comment]
I took your first point into account with 66483d9 - things are saved properly now.
[testing_mission_random_value.txt](https://github.com/endless-sky/endless-sky/files/908062/testing_mission_random_value.txt)

My coding intention was to avoid that if empty() by simply accepting that svalue only if it results in something useful. In the ::Test function you set the firstValue to zero first too, and only overwrite it if you have some sort of match, dont you?

What I take from you coding standard remark is the issue code duplication, correct?
I will try to work that out.

**Comment by EndrosG** [comment]
@endless-sky : Those two clean up commits should be what you were asking for in terms of coding standard.
The code builds.

The above supplied testing mission is saved to:
```
mission "TEST Random Value"
	name "Random stuff"
	landing
	to offer
		random < random
	to fail
		"reputation: Pirate" > "reputation: Merchant"
	destination Earth
	on offer
		conversation
			label 0
			`The landing officer welcomes you with a short "Hey Captain <last>, you have passed the test."`
				accept
		"new variable" = 20
		"random var" = random
```

**Comment by endless-sky** [comment]
As-is, this code reports error messages every time a condition with a non-numeric value is encountered. If you're not launching the program from the command line, you'll need to look at the "errors.txt" file in the configuration directory (one level up from the saved games) to find out if errors are being reported.

Also, this would interpret the number "0" as a string (because it's Value() is 0). That ends up not mattering unless the game defines a condition named "0" with a nonzero value, but it made more sense to me to add a more rigorous check for whether the value is a number or not (43034d6). The other changes in that commit are stylistic: `if(` instead of `if (`, brackets on their own lines, and functions in the .cpp listed in the same order as the .h.

**Comment by EndrosG** [comment]
That one space between the if and the bracket was a mistake, not intended. I know about that part of the coding standard. Sorry again.
You additional commit looks good! :+1:

File: source/ConditionSet.cpp [hunk 1]
```diff
@@ -40,7 +40,6 @@ namespace {
 			{"=", [](int a, int b) { return b; }},
 			{"+=", [](int a, int b) { return a + b; }},
 			{"-=", [](int a, int b) { return a - b; }},
-			{"*=", [](int a, int b) { return a * b; }},
```
Comment (tehhowch) [inline source/ConditionSet.cpp]: Was this removal intended?

**Comment by EndrosG** [comment]
I did "fix" this missing *= later in my "substitute conditions" branch since that was thought as followup but sadly never merged.

**Comment by tehhowch** [comment]
Seems like restoring it is a ready-to-merge standalone PR :)
You might also venture to define the `/=` operator:
```
#include <limits>

{"/=", [](int a, int b) { return b ? a * (1. / b) : numeric_limits<int>::max(); }
```
but as ConditionSets appear to deal only with integral returns, the fractional portions (which may be strongly desired / thought to be existent and used) are likely to be lost.


PS: welcome back :+1:
